examples/structured_data/deep_neural_decision_forests.py

```diff
@@ -0,0 +1,454 @@
+"""
+Title: Classification with Neural Decision Forests
+Author: [Khalid Salama](https://www.linkedin.com/in/khalid-salama-24403144/)
+Date created: 2021/01/15
+Last modified: 2021/01/15
+Description: Structured data classification with Deep Neural Decision Forests.
+"""
+
+"""
+## Introduction
+
+This example provides an implementation of the [Deep Neural Decision Forest](https://www.cv-foundation.org/openaccess/content_iccv_2015/papers/Kontschieder_Deep_Neural_Decision_ICCV_2015_paper.pdf)
+model, introduced by P. Kontschieder et al., for structured data classification.
+This model introduced a stochastic and differentiable decision tree model to unify
+classification trees with the deep representation learning, in a joint training routine.
+
+## The dataset
+
+This example uses the [United States Census Income
+Dataset](https://archive.ics.uci.edu/ml/datasets/census+income) provided by the
+[UC Irvine Machine Learning
+Repository](https://archive.ics.uci.edu/ml/index.php). The task is binary classification
+to determine whether a person makes over 50K a year.
+
+The dataset includes 48,842 instances with 14 input features, 5 of which numerical,
+and the other 9 are categorical.
+"""
+
+"""
+## Setup
+"""
+
+import tensorflow as tf
+import numpy as np
+import pandas as pd
+from tensorflow import keras
+from tensorflow.keras import layers
+import math
+
+"""
+## Prepare the data
+"""
+
+CSV_HEADER = [
+    "age",
+    "workclass",
+    "fnlwgt",
+    "education",
+    "education_num",
+    "marital_status",
+    "occupation",
+    "relationship",
+    "race",
+    "gender",
+    "capital_gain",
+    "capital_loss",
+    "hours_per_week",
+    "native_country",
+    "income_bracket",
+]
+
+train_data_url = (
+    "https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data"
+)
+train_data = pd.read_csv(train_data_url, header=None, names=CSV_HEADER)
+
+test_data_url = (
+    "https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.test"
+)
+test_data = pd.read_csv(test_data_url, header=None, names=CSV_HEADER)
+
+print(f"Train dataset shape: {train_data.shape}")
+print(f"Test dataset shape: {test_data.shape}")
+
+train_data.head().T
+
+"""
+We remove the first record as it is not a valid data example, and we remove a tailing
+'dot' in the class labels.
+"""
+
+test_data = test_data[1:]
+test_data.income_bracket = test_data.income_bracket.apply(
+    lambda value: value.replace(".", "")
+)
+
+"""
+Now we store the train and test data splits locally to CSV files.
+"""
+
+train_data_file = "train_data.csv"
+test_data_file = "test_data.csv"
+
+train_data.to_csv(train_data_file, index=False, header=False)
+test_data.to_csv(test_data_file, index=False, header=False)
+
+"""
+## Define dataset metadata
+
+Here, we define the metadata of the dataset that will be useful for reading and parsing
+the data into input features, and encoding the input features with respect to their types.
+"""
+
+NUMERIC_FEATURE_NAMES = [
```

Comment:
Add a comment above each global variable you define

Reply:
Done